scripts/metricsgen/metricsdiff/metricsdiff.go

```diff
@@ -0,0 +1,156 @@
+// containing prometheus metrics output.
+type MetricsDiff struct {
+	Adds    []string
+	Removes []string
+
+	LabelDiffs []LabelDiff
+}
+
+// MetricsDiff contains the set of labels that were modified for a single metric
+// between two files containing prometheus metrics output.
+type LabelDiff struct {
+	MetricsName string
+	Adds        []string
+	Removes     []string
+}
+
+func main() {
+	flag.Parse()
+	if len(os.Args) != 3 {
+		log.Fatalf("Incorrect number of arguments, expected 2 but received %d", len(os.Args)-1)
+	}
+	fa, err := os.Open(os.Args[1])
+	if err != nil {
+		log.Fatalf("Unable to open file %s: %v", os.Args[1], err)
+	}
+	fb, err := os.Open(os.Args[2])
+	if err != nil {
+		log.Fatalf("Unable to open file %s: %v", os.Args[2], err)
+	}
+	md, err := MetricsDiffFromReaders(fa, fb)
+	if err != nil {
+		log.Fatalf("Generating diff: %v", err)
+	}
+	fmt.Printf("%s", md)
+}
+
+// MetricsDiffFromReaders parses the metrics present in the readers a and b and
+// determines which metrics were added and removed in b.
+func MetricsDiffFromReaders(a, b io.Reader) (MetricsDiff, error) {
+	parser := expfmt.TextParser{}
+	amf, err := parser.TextToMetricFamilies(a)
+	if err != nil {
+		return MetricsDiff{}, err
+	}
+	bmf, err := parser.TextToMetricFamilies(b)
+	if err != nil {
+		return MetricsDiff{}, err
+	}
+
+	md := MetricsDiff{}
+	for name, afamily := range amf {
+		bfamily, ok := bmf[name]
+		if !ok {
+			md.Removes = append(md.Removes, name)
+			continue
+		}
+
+		labelsDiff := false
+		aLabelSet := toSet(afamily.Metric[0].Label)
+		bLabelSet := toSet(bfamily.Metric[0].Label)
+		ld := LabelDiff{
+			MetricsName: name,
+		}
+		for name := range aLabelSet {
+			_, ok := bLabelSet[name]
+			if !ok {
+				ld.Removes = append(ld.Removes, name)
+				labelsDiff = true
+			}
+		}
+		for name := range bLabelSet {
+			_, ok := aLabelSet[name]
+			if !ok {
+				ld.Adds = append(ld.Adds, name)
+				labelsDiff = true
+			}
+		}
+		if labelsDiff {
+			md.LabelDiffs = append(md.LabelDiffs, ld)
+		}
+	}
+	for name := range bmf {
+		if _, ok := amf[name]; !ok {
+			md.Adds = append(md.Adds, name)
+		}
+	}
+	return md, nil
+}
+
+func toSet(lps []*dto.LabelPair) map[string]struct{} {
+	m := make(map[string]struct{}, len(lps))
+	for _, pair := range lps {
+		m[*pair.Name] = struct{}{}
+	}
+	return m
+}
+
+func (m MetricsDiff) String() string {
+	var s string
+	if len(m.Adds) > 0 {
+		s += "Adds: \n"
+		for _, add := range m.Adds {
+			s += fmt.Sprintf("+++ %s\n", add)
+		}
+	}
+	if len(m.Removes) > 0 {
+		s += "Removes: \n"
+		for _, remove := range m.Removes {
+			s += fmt.Sprintf("--- %s\n", remove)
+		}
+	}
+	if len(m.LabelDiffs) > 0 {
+		s += "Label Changes: \n"
+		for _, ld := range m.LabelDiffs {
+			s += fmt.Sprintf("Label: %s\n", ld.MetricsName)
+			for _, add := range ld.Adds {
+				s += fmt.Sprintf("+++ %s", add)
+			}
+			for _, remove := range ld.Removes {
+				s += fmt.Sprintf("--- %s", remove)
+			}
+		}
+	}
+	return s
```

Comment:
Not that memory matters here, but a suggestion:
```suggestion
	var s strings.Builder
	if len(m.Adds) > 0 {
		fmt.Fprintln(&s, "Adds:")
		for _, add := range m.Adds {
			fmt.Fprintf(&s, "+++ %s\n", add)
		}
	}
	if len(m.Removes) > 0 {
		fmt.Fprintln(&s, "Removes:")
		for _, rem := range m.Removes {
			fmt.Fprintf(&s, "--- %s\n", rem)
		}
	}
	if len(m.LabelDiffs) > 0 {
		fmt.Fprintln(&s, "Label changes:")
		for _, ld := range m.LabelDiffs {
			fmt.Fprintf(&s, "Label: %s\n", ld.MetricsName)
			for _, add := range ld.Adds {
				fmt.Fprintf(&s, "+++ %s\n", add)
			}
			for _, rem := range ld.Removes {
				fmt.Fprintf(&s, "--- %s\n", rem)
			}
		}
	}
	return s.String()
```

Reply:
I thought about this but was taking a 'let's get this on its feet asap' approach. The suggested change here is good.